File: java/src/test/java/com/example/routeauthorizationtesting/WebSecurityConfigTest.java
package com.example.routeauthorizationtesting;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.mvc.method.RequestMappingInfo;
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;

import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.DynamicTest.dynamicTest;

@SpringBootTest
@AutoConfigureMockMvc
class WebSecurityConfigTest {
    final Set<RouteAuthSpec> routeAuthSpecs = Set.of(
        new RouteAuthSpec("/user/{id}", HttpMethod.GET, new Access.Unauthenticated()),
        new RouteAuthSpec("/user/{id}", HttpMethod.PUT, new Access.AnyRole(Role.ADMIN, Role.BASIC)),
        new RouteAuthSpec("/user/{id}", HttpMethod.PATCH, new Access.AnyRole(Role.ADMIN, Role.BASIC)),
        new RouteAuthSpec("/user", HttpMethod.POST, new Access.AnyRole(Role.ADMIN)),
        new RouteAuthSpec("/user/{id}", HttpMethod.DELETE, new Access.AnyRole(Role.ADMIN)),
        new RouteAuthSpec("/some-route-without-correct-auth-to-show-failure", HttpMethod.POST, new Access.AnyRole(Role.ADMIN)),
        new RouteAuthSpec("/some-route-not-implemented-to-show-failure", HttpMethod.GET, new Access.Unauthenticated())
    );

    @Autowired
    MockMvc mockMvc;

    @Autowired
    RequestMappingHandlerMapping requestMapping;

    @Test
    void testsEveryRouteAndNoNonExistentRoutes() {
        Set<String> allRoutes = requestMapping.getHandlerMethods().keySet().stream()
            .filter((RequestMappingInfo info) ->
                info.getPatternsCondition() != null &&
                    !info.getPatternsCondition().getPatterns().contains("/error")
            ).flatMap((RequestMappingInfo info) -> {
                RequestMethod method = info.getMethodsCondition().getMethods().iterator().next();
                Set<String> routes = info.getPatternsCondition().getPatterns();
                return routes.stream().map((r) -> String.format("%s %s", method, r));
            }).collect(Collectors.toSet());

        Set<String> testedRoutes = routeAuthSpecs.stream()
            .map((s) -> String.format("%s %s", s.verb, s.route))
            .collect(Collectors.toSet());

        Set<String> untestedRoutes = allRoutes.stream()
            .filter((r) -> !testedRoutes.contains(r))
            .collect(Collectors.toSet());

        Set<String> nonexistentRoutes = testedRoutes.stream()
            .filter((r) -> !allRoutes.contains(r))
            .collect(Collectors.toSet());

        assertAll(
            () -> assertTrue(untestedRoutes.isEmpty(), String.format("The following routes are untested: %s", untestedRoutes)),
            () -> assertTrue(nonexistentRoutes.isEmpty(), String.format("Tests are defined for the following nonexistent routes: %s", nonexistentRoutes))
        );
    }

    @TestFactory
    List<DynamicTest> allRoutesHaveCorrectAuthorizationWhenUnauthenticated() {
        return routeAuthSpecs.stream().map((spec) -> dynamicTest(
            String.format("Unauthenticated [%s] %s", spec.verb, spec.route), () -> {
                assertCorrectUnauthenticated(spec);
            }
        )).collect(Collectors.toList());
    }

    @TestFactory
    List<DynamicTest> allRoutesHaveCorrectAuthorizationWhenAuthenticated() {
        return Arrays.stream(Role.values()).flatMap((role) ->
            routeAuthSpecs.stream().map((spec) -> dynamicTest(
                String.format("%s [%s] %s", role, spec.verb, spec.route), () -> {
                    assertCorrectAuthz(role, spec);
                }
            ))
        ).collect(Collectors.toList());
    }

    void assertCorrectUnauthenticated(RouteAuthSpec spec) throws Exception {
        MvcResult result = mockMvc.perform(spec.getRequest()).andReturn();

        assertNotEquals(HttpStatus.METHOD_NOT_ALLOWED.value(), result.getResponse().getStatus(),
            String.format("Method %s does not exist for route %s", spec.verb, spec.route));

        if (spec.access instanceof Access.Unauthenticated) {
            assertNotEquals(HttpStatus.UNAUTHORIZED.value(), result.getResponse().getStatus(),
                String.format("Expected %s %s not to require authentication", spec.verb, spec.route));
        } else {
            assertEquals(HttpStatus.UNAUTHORIZED.value(), result.getResponse().getStatus(),
                String.format("Expected %s %s to require authentication.", spec.verb, spec.route));
        }
    }


    void assertCorrectAuthz(Role role, RouteAuthSpec spec) throws Exception {
        MvcResult result = mockMvc.perform(
            spec.getRequest().with(
                SecurityMockMvcRequestPostProcessors
                    .user("automation-user")
                    .roles(role.toString())
            )
        ).andReturn();

        if (spec.access instanceof Access.AnyRole) {
            if (((Access.AnyRole) spec.access).allowedForRole(role)) {
                assertNotEquals(HttpStatus.FORBIDDEN.value(), result.getResponse().getStatus(),
                    String.format("Expected role %s to be PERMITTED to %s %s", role, spec.verb, spec.route));
            } else {
                assertEquals(HttpStatus.FORBIDDEN.value(), result.getResponse().getStatus(),
                    String.format("Expected role %s to be FORBIDDEN to %s %s", role, spec.verb, spec.route));
            }
        } else if (spec.access instanceof Access.Unauthenticated) {
            assertNotEquals(HttpStatus.FORBIDDEN.value(), result.getResponse().getStatus(),
                String.format("Expected role %s to be PERMITTED to %s %s", role, spec.verb, spec.route));
        } else {
            throw new Error("Unknown class passed for spec.access");
        }
    }
}

File: kotlin/src/test/kotlin/com/example/routeauthorizationtesting/WebSecurityConfigTest.kt
package com.example.routeauthorizationtesting

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.test.web.servlet.MockMvc
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping

@SpringBootTest
@AutoConfigureMockMvc
internal class WebSecurityConfigTest {

    val routeAuthSpecs: Set<RouteAuthSpec> = setOf(
        RouteAuthSpec("/user/{id}", HttpMethod.GET, Access.Unauthenticated),
        RouteAuthSpec("/user/{id}", HttpMethod.PUT, Access.AnyRole(Role.ADMIN, Role.BASIC)),
        RouteAuthSpec("/user/{id}", HttpMethod.PATCH, Access.AnyRole(Role.ADMIN, Role.BASIC)),
        RouteAuthSpec("/user", HttpMethod.POST, Access.AnyRole(Role.ADMIN)),
        RouteAuthSpec("/user/{id}", HttpMethod.DELETE, Access.AnyRole(Role.ADMIN)),
        RouteAuthSpec("/some-route-without-correct-auth-to-show-failure", HttpMethod.POST, Access.AnyRole(Role.ADMIN)),
        RouteAuthSpec("/some-route-not-implemented-to-show-failure", HttpMethod.GET, Access.Unauthenticated),
    )

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var requestMapping: RequestMappingHandlerMapping

    @Test
    fun `tests every route and no non-existent routes`() {
        val allRoutes = requestMapping.handlerMethods.keys
            .filter { it.patternsCondition != null }
            .filterNot { it.patternsCondition!!.patterns.contains("/error") }
            .flatMap { requestMappingInfo ->
                val method = requestMappingInfo.methodsCondition.methods.first()
                val routes = requestMappingInfo.patternsCondition!!.patterns
                routes.map { route -> "$method $route" }
            }
        val testedRoutes = routeAuthSpecs.map { spec ->
            "${spec.verb} ${spec.route}"
        }

        val untestedRoutes: Set<String> = allRoutes.subtract(testedRoutes)
        val nonexistentRoutes = testedRoutes.minus("GET /index").subtract(allRoutes)

        assertAll({
            assertThat(untestedRoutes)
                .withFailMessage("The following routes are untested: %s", untestedRoutes)
                .isEmpty()
        }, {
            assertThat(nonexistentRoutes)
                .withFailMessage("Tests are defined for the following nonexistent routes: %s", nonexistentRoutes)
                .isEmpty()
        })
    }


    @TestFactory
    fun `all routes have correct authorization when authenticated`(): List<DynamicTest> =
        Role.values().flatMap { role ->
            routeAuthSpecs.map { spec ->
                dynamicTest("$role [${spec.verb}] ${spec.route}") {
                    assertCorrectAuthz(role, spec)
                }
            }
        }

    @TestFactory
    fun `all routes have correct authorization when unauthenticated`(): List<DynamicTest> =
        routeAuthSpecs.map { spec ->
            dynamicTest("Unauthenticated [${spec.verb}] ${spec.route}") {
                assertCorrectUnauthenticated(spec)
            }
        }

    private fun assertCorrectAuthz(role: Role, spec: RouteAuthSpec) {
        val result = mockMvc.perform(
            spec.request.with(
                SecurityMockMvcRequestPostProcessors
                    .user("automation-user")
                    .roles(role.toString())
            )
        ).andReturn()

        when (spec.access) {
            is Access.AnyRole -> {
                if (spec.access.allowedForRole(role)) {
                    assertThat(result.response.status)
                        .withFailMessage("Expected role $role to be PERMITTED to ${spec.verb} ${spec.route}")
                        .isNotEqualTo(HttpStatus.FORBIDDEN.value())
                } else {
                    assertThat(result.response.status)
                        .withFailMessage("Expected role $role to be FORBIDDEN to ${spec.verb} ${spec.route}")
                        .isEqualTo(HttpStatus.FORBIDDEN.value())
                }
            }
            is Access.Unauthenticated -> {
                assertThat(result.response.status)
                    .withFailMessage("Expected role $role to be PERMITTED to ${spec.verb} ${spec.route}")
                    .isNotEqualTo(HttpStatus.FORBIDDEN.value())
            }
        }
    }

    private fun assertCorrectUnauthenticated(spec: RouteAuthSpec) {
        val result = mockMvc.perform(spec.request).andReturn()

        assertThat(result.response.status)
            .withFailMessage("Method ${spec.verb} does not exist for route ${spec.route}")
            .isNotEqualTo(HttpStatus.METHOD_NOT_ALLOWED.value())

        when (spec.access) {
            is Access.Unauthenticated -> {
                assertThat(result.response.status)
                    .withFailMessage("Expected ${spec.verb} ${spec.route} not to require authentication")
                    .isNotEqualTo(HttpStatus.UNAUTHORIZED.value())
            }
            else -> {
                assertThat(result.response.status)
                    .withFailMessage("Expected ${spec.verb} ${spec.route} to require authentication.")
                    .isEqualTo(HttpStatus.UNAUTHORIZED.value())
            }
        }
    }
}